feat!(KeyBindings): Allow pure modifier keysets

badosu opened the pull request:
Previously the engine was unable to handle any combination of modifiers without keysets, for example:

An input of keypress ctrl+alt -> would trigger any action bound to alt (engine changes the internal representation to `any+<original_keyset>`, in this example `any+alt`: as `ctrl` is a modifier it gets gobbled by the `any`)

This commit makes it so that any keyboard state composed of a combination of modifiers without any other keys is interpreted as a keyset of 'none' + mods.

For consumers intending to bind to these combinations, none is not required as we sanitize internally, but it can also be set, e.g.:

Shift+none == shift  or  ctrl+shift == ctrl+shift+none

Fix #1071

badosu committed:
feat!(KeyBindings): Allow pure modifier keysets

Previously the engine was unable to handle any combination of modifiers
without keysets, for example:

An input of keypress alt -> would trigger an action bound to ctrl+alt

This commit makes it so that any keyboard state composed of a
combination of modifiers without any other keys is interpreted as a
keyset of 'none' + mods.

For consumers intending to bind to these combinations, none is not
required as we sanitize internally, but it can also be set, e.g.:

Shift+none == shift  or  ctrl+shift == ctrl+shift+none

badosu commented on rts/Game/UI/KeySet.cpp:
```diff
@@ -193,11 +211,11 @@ bool CKeySet::Parse(const std::string& token, bool showerror)
 		return false;
 	}
 
-	if (IsModifier())
```
This made absolutely no sense, I have no idea why this was ever a case.

Perhaps a previous 'lets be safe' case originating from statefulCommands handling like in KeyBindings.cpp (that still remains the case) that didn't get removed.

sprunk replied:
Is the behaviour for non-modifier-only bindings with Any still the same? Say things like `Any+Ctrl+X`.

badosu replied:
This is a more "no idea why this exists, there's not a reasonable explanation for this" than "no idea why this exists, I don't understand what this does, lets remove it".

The only cases that will work differently are ones that rely on pure modifier bindings, e.g. `bind something [mod\+*]?mod`, the difference being that they wont be equivalent to `any+mod` anymore, which is an improvement.

No games rely on this unexpected behavior as far as I have investigated.

badosu committed:
hue

badosu committed:
Merge branch 'BAR105' of github.com:beyond-all-reason/spring into badosu/allow-pure-modifiers-keysets

rhys-vdw commented on rts/Game/UI/KeyBindings.cpp:
```diff
@@ -620,12 +631,16 @@ void CKeyBindings::AddActionToKeyMap(KeyMap& bindings, Action& action)
-			al.push_back(action);
-		}
+		if (it != std::end(al)) {
+      return false;
```
Mixing spaces and tabs here.

badosu committed:
Merge branch 'master' of github.com:beyond-all-reason/spring into badosu/allow-pure-modifiers-keysets

badosu committed:
Fix invalid return value

badosu committed:
Fix /unbindaction not unbinding from scancodes in some cases

When an action was bound by both scancodes and keycodes, it would unbind only from
keycode bindings on /unbindaction

badosu committed:
Fix tabs

badosu committed:
Rebuild actionlists when fakemeta is set

badosu committed:
Stuff works, lets not bother with fakemeta resets for now
